Add best practices for HTTP 301 and retries to Spark vignette

mweiden opened the pull request:
When querying the DSS API, clients are expected
* Handle HTTP 301 responses and retry their requests after the number of seconds specified in the `retry-after` header
* Retry their requests to the API as requests can fail when the api is under high load

This PR demonstrates these best practices in the data consumer vignettes.

Fixes https://github.com/HumanCellAtlas/data-store/issues/1263

briandoconnor commented on tasks/SPARK/src/main/java/kmer/Kmer.java:
```diff
@@ -0,0 +1,219 @@
+ * sizes are referred to as 12-mers, 20-mers, and so forth.
+ *
+ * @author Mahmoud Parsian
+ * @author Brian O'Connor (made tweaks for HCA demo)
```
@mweiden you should add yourself as an author too, right?

briandoconnor commented on tasks/SPARK/src/test/java/kmer/DataStoreTest.java:
```diff
@@ -0,0 +1,63 @@
+package kmer;
```
So happy to see you added tests!!!

briandoconnor commented on tasks/SPARK/src/main/java/kmer/HttpUtil.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    public static JSONObject jsonRequest(String url) throws IOException, InterruptedException {
+        Optional<JSONObject> maybeJSON = Retriable.runWithRetries(
```
Great!!!

briandoconnor commented on tasks/SPARK/src/test/java/kmer/RetriableTest.java:
```diff
@@ -0,0 +1,38 @@
+    @Test
+    public void runWithRetriesShouldRetry() throws InterruptedException {
+        TestRetriable retriable = new TestRetriable();
+        Retriable.runWithRetries( 3, 0, retriable);
```
Do you want to try a retry test that uses a mock web server to return a 503 (that's what the blue box is using to signal a try again later, right?)

briandoconnor commented on tasks/SPARK/src/main/java/kmer/Retriable.java:
```diff
@@ -0,0 +1,25 @@
+    static <T> Optional<T> runWithRetries(int maxRetries, int maxInterval, Retriable<T> t) throws InterruptedException {
+        int count = 0;
+        while (count < maxRetries) {
+            try {
```
Does a 503 really cause an exception?  Does this need to be tested through a unit test?

mweiden replied:
Yep. Very Java.

mweiden commented:
@briandoconnor I addressed your comments and made some other improvements including command line arg configuration with flags. I've run this successfully against the prod deployment. Merging this now.
